File: group 4/variant07/ParcelDeliveryService.java
public final class ParcelDeliveryService {
    public record Request(int weightGrams, int distanceKm, boolean express, boolean insured) {}

    public record Result(String status, int billableKilograms, long basePrice, long distanceFee, long optionsFee, long total) {}

    public interface RouteCapacity {
        int availableGrams();
    }

    public interface ExpressPolicy {
        boolean isSupported();
    }

    public interface ShipmentRepository {
        void save(Request request, Result result);
    }

    private final RouteCapacity capacity;
    private final ExpressPolicy express;
    private final ShipmentRepository shipments;

    public ParcelDeliveryService(RouteCapacity capacity, ExpressPolicy express, ShipmentRepository shipments) {
        this.capacity = java.util.Objects.requireNonNull(capacity, "Залежність не задано: capacity");
        this.express = java.util.Objects.requireNonNull(express, "Залежність не задано: express");
        this.shipments = java.util.Objects.requireNonNull(shipments, "Залежність не задано: shipments");
    }

    public Result process(Request request) {
        if (request == null) {
            throw new IllegalArgumentException("Заявку не задано");
        }
        checkRange(request.weightGrams(), 1, 29_999, "Вага посилки");
        checkRange(request.distanceKm(), 1, 3_000, "Відстань");
        int available = capacity.availableGrams();
        if (available < 0) {
            throw new IllegalStateException("Від’ємна доступна вантажність");
        }
        if (available < request.weightGrams()) {
            return rejected("Недостатня вантажність");
        }
        if (request.express() && !express.isSupported()) {
            return rejected("Експрес-доставка недоступна");
        }
        int kilograms = (request.weightGrams() + 999) / 1_000;
        long basePrice = 4_000 + kilograms * 1_000L;
        long distanceFee = request.distanceKm() >= 500 ? 3_000 : 0;
        long options;
        if (request.express() && request.insured()) {
            options = 6_000;
        } else if (request.express()) {
            options = 4_000;
        } else if (request.insured()) {
            options = 2_000;
        } else {
            options = 0;
        }
        Result result = new Result("Прийнято", kilograms, basePrice, distanceFee, options,
                basePrice + distanceFee + options);
        shipments.save(request, result);
        return result;
    }

    private static Result rejected(String status) {
        return new Result(status, 0, 0, 0, 0, 0);
    }

    private static void checkRange(int value, int minimum, int maximum, String field) {
        if (value < minimum || value > maximum) {
            throw new IllegalArgumentException("Недопустиме значення: " + field);
        }
    }
}

File: group 2/variant07/parcel_delivery_service.py
from dataclasses import dataclass
from typing import Protocol


@dataclass(frozen=True)
class Request:
    weight_grams: int
    distance_km: int
    express: bool
    insured: bool


@dataclass(frozen=True)
class Result:
    status: str
    billable_kilograms: int = 0
    base_price: int = 0
    distance_fee: int = 0
    options_fee: int = 0
    total: int = 0


class RouteCapacity(Protocol):
    def available_grams(self) -> int:
        ...


class ExpressPolicy(Protocol):
    def is_supported(self) -> bool:
        ...


class ShipmentRepository(Protocol):
    def save(self, request: Request, result: Result) -> None:
        ...


def _range(value: int, minimum: int, maximum: int, field: str) -> None:
    if type(value) is not int or not minimum <= value <= maximum:
        raise ValueError("Недопустиме значення: " + field)


def _flag(value: bool, field: str) -> bool:
    if type(value) is not bool:
        raise RuntimeError("Некоректні відомості: " + field)
    return value


def _count(value: int, field: str) -> int:
    if type(value) is not int or value < 0:
        raise RuntimeError("Некоректні відомості: " + field)
    return value


class ParcelDeliveryService:
    def __init__(self, capacity: RouteCapacity, express: ExpressPolicy, shipments: ShipmentRepository) -> None:
        if capacity is None or express is None or shipments is None:
            raise ValueError("Усі залежності мають бути задані")
        self._capacity = capacity
        self._express = express
        self._shipments = shipments

    def process(self, request: Request) -> Result:
        if not isinstance(request, Request):
            raise ValueError("Заявку не задано або її формат некоректний")
        _range(request.weight_grams, 1, 29_999, "Вага посилки")
        _range(request.distance_km, 1, 3_000, "Відстань")
        if type(request.express) is not bool or type(request.insured) is not bool:
            raise ValueError("Ознаки мають бути логічними значеннями")
        capacity = _count(self._capacity.available_grams(), "Доступна вантажність")
        if capacity < request.weight_grams:
            return Result("Недостатня вантажність")
        if request.express and not _flag(self._express.is_supported(), "Доступність експрес-доставки"):
            return Result("Експрес-доставка недоступна")
        kilograms = (request.weight_grams + 999) // 1_000
        base_price = 4_000 + kilograms * 1_000
        distance_fee = 3_000 if request.distance_km >= 500 else 0
        if request.express and request.insured:
            options = 6_000
        elif request.express:
            options = 4_000
        elif request.insured:
            options = 2_000
        else:
            options = 0
        result = Result("Прийнято", kilograms, base_price, distance_fee, options,
                        base_price + distance_fee + options)
        self._shipments.save(request, result)
        return result
